### PR TITLE
Tag related cases as such in case search results

### PR DESCRIPTION
**Pull request by esoergel**
## Summary
Accompanies https://github.com/dimagi/commcare-hq/pull/30323
from https://dimagi-dev.atlassian.net/browse/USH-1252
First two commits are cherry-picked from a WIP branch I have locally, but make these tests easier to set up.

## Feature Flag
Case search & claim

## Product Description
Prevents invalid results from showing up in case search. Unlikely to be relevant for non-USH projects.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

The components are tested individually, this adds tests for the piece as a whole

### QA Plan

N/A

### Safety story

This is reasonably well-tested, and doesn't actually change much

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
Did a local test of the entire workflow on a branch containing this plus https://github.com/dimagi/commcare-hq/pull/30323, worked fine.

**Comment by orangejenny** [comment]
Test failure looks related: 
```
ERROR: corehq.apps.es.tests.test_case_search_es:TestCaseSearchLookups.test_get_related_case_results
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/es/tests/test_case_search_es.py", line 507, in test_get_related_case_results
    self._assert_related_case_ids(cases, {"parent/parent"}, {"c1"})
  File "/mnt/commcare-hq-ro/corehq/apps/es/tests/test_case_search_es.py", line 513, in _assert_related_case_ids
    results = get_related_case_results(self.domain, cases, paths)
  File "/mnt/commcare-hq-ro/corehq/apps/case_search/utils.py", line 266, in get_related_case_results
    indices = [case.get_index(identifier) for case in current_cases]
  File "/mnt/commcare-hq-ro/corehq/apps/case_search/utils.py", line 266, in <listcomp>
    indices = [case.get_index(identifier) for case in current_cases]
AttributeError: 'dict' object has no attribute 'get_index'
```

**Comment by esoergel** [comment]
Yeah you're right - I reverted the last commit, which should fix it.

File: corehq/apps/es/tests/utils.py [hunk 1]
```diff
@@ -61,10 +61,18 @@ def es_test(test):
     return attr(es_test=True)(test)
 
 
-def populate_user_index(user_docs):
+def populate_es_index(models, index_name, doc_prep_fn=lambda doc: doc):
+    index_info = ES_META[index_name]
     es = get_es_new()
-    ensure_index_deleted(USER_INDEX)
```
Comment (snopoke) [inline corehq/apps/es/tests/utils.py:66]: is it ok to remove this line?

File: corehq/apps/es/tests/utils.py [hunk 2]
```diff
@@ -76,3 +79,18 @@ def populate_es_index(models, index_name, doc_prep_fn=lambda doc: doc):
...
 def populate_user_index(users):
     populate_es_index(users, 'users')
+
+
+def case_search_es_setup(domain, case_blocks):
+    """Submits caseblocks, creating the cases, then sends them to ES"""
+    from corehq.apps.hqcase.utils import submit_case_blocks
+    _, cases = submit_case_blocks([cb.as_text() for cb in case_blocks], domain=domain)
+    order = {cb.case_id: index for index, cb in enumerate(case_blocks)}
+    # send cases to ES in the same order they were passed in
```
Comment (snopoke) [inline corehq/apps/es/tests/utils.py:90]: Why is this important? I see in the other comment it says it's to get the `indexed_on` field to match the creation order by why do we care about that?

File: corehq/apps/case_search/utils.py [hunk 3]
```diff
@@ -13,8 +22,41 @@
     FuzzyProperties,
 )
 from corehq.apps.es.case_search import CaseSearchES, flatten_result
-from corehq.apps.case_search.const import CASE_SEARCH_MAX_RESULTS
-from corehq.apps.case_search.filter_dsl import CaseFilterError
+
+
+def get_case_search_results(domain, criteria, app_id=None):
```
Comment (snopoke) [inline corehq/apps/case_search/utils.py:27]: :+1:

**Review by snopoke — APPROVED** [review]
Few small comments - nothing blocking

File: corehq/apps/es/tests/utils.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/es/tests/utils.py:66]: It should be - this is used in two places and both also include this in their teardown.  The risk is there's someplace else that doesn't, but if the test suite passes I assume it's fine.

File: corehq/apps/es/tests/utils.py (hunk 2 above)
Comment (esoergel) [reply corehq/apps/es/tests/utils.py:90]: It was so the results would be returned in a predictable order for tests (`submit_case_blocks` doesn't preserve order).  This is used in the test suite for the case list API, which paginates by `indexed_on` - this way I know which cases fall on which side of each page break.  Do you think that's leaking implementation across the abstraction boundary?

Comment (snopoke) [reply corehq/apps/es/tests/utils.py:90]: I think it's OK but maybe just update the comment to say 'why'

**Comment by esoergel** [comment]
Just rebased to trigger a travis build - looks like it didn't start one after the revert.  No new changes.

**Pull request merged**